gridcomps/ExtData3G/AbstractDataSetFileSelector.F90

```diff
@@ -0,0 +1,148 @@
+       if (file_found) then
+          filename = trial_file
+          _RETURN(_SUCCESS)
+       end if
+       do i=1, MAX_TRIALS
+          useable_time = useable_time + this%frequency
+          call fill_grads_template(trial_file, this%file_template, time=useable_time, _RC)
+          inquire(file=trim(trial_file),exist=file_found)
+          if (file_found) then
+             filename = trial_file
+             exit
+          end if
+       enddo
+       _RETURN(_SUCCESS)
+
+    end function find_any_file
+
+    function compute_trial_time(this, target_time, shift, rc) result(trial_time)
+       type(ESMF_Time) :: trial_time
+       class(AbstractDataSetFileSelector), intent(inout) :: this
+       type(ESMF_Time), intent(in) :: target_time
+       integer, intent(in) :: shift
+       integer, optional, intent(out) :: rc
+
+       integer :: status, n
+       integer(ESMF_KIND_I8) :: int_sec
+       call ESMF_TimeIntervalGet(this%frequency, s_i8=int_sec, _RC)
+       if (int_sec == 0) then
+          trial_time = this%ref_time
+          do while(trial_time <= target_time)
+             trial_time = trial_time + this%frequency
+        
+          enddo
+          trial_time = trial_time - this%frequency + shift*this%frequency
+       else
+          n = (target_time-this%ref_time)/this%frequency
+          trial_time = this%ref_time+(n+shift)*this%frequency
+       end if
+       _RETURN(_SUCCESS)
+       
+    end function compute_trial_time 
+
+    subroutine set_last_update(this, update_time, rc)
+       class(AbstractDataSetFileSelector), intent(inout) :: this
+       type(ESMF_Time), intent(in) :: update_time
+       integer, optional, intent(out) :: rc
+
+       integer :: status
+       if (allocated(this%last_updated)) then
+          this%last_updated = update_time
+       else
+          allocate(this%last_updated, source=update_time, _STAT)
+       end if
+       _RETURN(_SUCCESS)
+    end subroutine
+
+    function get_last_update(this) result(last_update)
+       type(ESMF_Time), allocatable :: last_update
+       class(AbstractDataSetFileSelector), intent(inout) :: this
+       if (allocated(this%last_updated)) last_update = this%last_updated 
+    end function
+    
+    function detect_time_flow(this, current_time, rc) result(time_jumped)
+       logical :: time_jumped
+       class(AbstractDataSetFileSelector), intent(inout) :: this
+       type(ESMF_Time), intent(in) :: current_time
+       integer, optional, intent(inout) :: rc
+
+       integer :: status
+       type(ESMF_TimeInterval) :: time_interval
+       integer(ESMF_KIND_I8) :: f1, f2
+
+       time_jumped = .false.
+       if (allocated(this%last_updated) .and. allocated(this%clock_dt)) then
+          time_interval = current_time - this%last_updated
+          call ESMF_TimeIntervalGet(time_interval, s_i8=f1, _RC)
+          call ESMF_TimeIntervalGet(this%clock_dt, s_i8=f2, _RC)
+          time_jumped = f2 < abs(f1)
```

Comment:
Somehow I find it easier to understand the intent by stating it as the change is larger than the clock dt rather than the other way around.   (Not at all sure if it should be `>` or `>=`.  So check this if you accept it.)
```suggestion
          time_jumped = abs(f1) > f2
```